## DSBD Project 2 -- LendingClub Peer-to-Peer Lending Prediction
### Intermediate Report

#### Yunhan (Claire) Xu, Qiyang (Cathy) Chen

Starting from 2015, the idea of Peer-to-Peer (P2P) lending emerged, which allows individuals (borrowers) to obtain loans directly from other individuals (investors) through platforms such as LendingClub, Prosper, and Upstart. This kind of "social lending" becomes an alternative financing approach, cutting out the financial institution as the middleman. P2P lending enables investors to acquire a higher return on investment compared to a bank deposit. However, credit risk, one of the biggest concerns for investors, emerges when they make informed investment decisions within the P2P lending procedure. In particular, a default due to failing to make the required payment within the agreed date for a borrower would induce a loss for an investor. 

In this project, we plan to develop a model that will predict whether a LendingClub-approved loan will end up being defaulted or not. The collected dataset includes loan grade, indicating the credit risk from the credit report and loan application. We plan to set the loan grade as the baseline and compare it with our model performance. In particular, we will draw a Precision-Recall Curve for the top 1%, 2%, 3%,..., 100% loans with default probability from high to low. We will also draw a Precision-Recall Curve for loan grade from A to G, and compare it with our models' Precision-Recall Curve. 

The domain of our project is as follows:
- Task (T) - Classification task that predicts defualted loan 
- Performance Measure (P) 
    - Precision: among all the actually defaulted loans, how many of them did we successfully identified
    - Recall: among all the loans that we predicted as default loans, how many of them actually defaulted
    - F-1 Score: weighted average for precision and recall
- Experience (E) - LendingClub database of loan records with loan status that we can use to label a loan as defaulted or not
<!-- machine-learned classification models (SVM, logistic regression, decision trees, random forests, XGBoost, etc.) -->

In sum, considering a classification task of identifying borrowers who are likely to make defaulted loans in LendingClub, we will train the LendingClub database of loan records without the pre-specified loan grades to perform the task and utilize precision, recall, and F-1 score as performance metrics to measure how well we complete the task.  

## Data

The original data contains all the LendingClub loan data from 2007 to 2018, and the corresponding 151 features are related to borrowers' credit history and loan characteristics. We performed some basic data cleaning beforehand:
- For the scoping purpose, we only focus on data from the latest four years, i.e., 2015-2018. 
- There are seven loan statuses: Charged Off, Current, Default, Fully Paid, In Grace Period, Late (16-30 days), Late (31-120 days). As we don't know whether a "Current" loan would be defaulted or not, we disregard those loans. We consider Late (16-30 days), Late (31-120 days), Default, and Charged Off as a Defaulted loan and Fully Paid as a desirable loan.

### Columns we discarded
- 100% missing values: `id`, `member_id`, `url`, `next_pymnt_d`
- The same for all rows: `pymnt_plan`, `out_prncp`, `out_prncp_inv`, and `policy_code`
- Updated in past 2 months -- wouldn't have these values at loan origination: `num_tl_120dpd_2m`, and `num_tl_30dpd`
- Hardship-related variables, relating to borrowers who were on a hardship plan -- wouldn't know at the point of loan origination whether or not the borrower will have a hardship: `deferral_term`, `hardship_amount`, `hardship_dpd`, `hardship_end_date`, `hardship_flag`, `hardship_last_payment_amount`, `hardship_length`, `hardship_loan_status`, `hardship_payoff_balance_amount`, `hardship_reason`, `hardship_start_date`, `hardship_status`, `hardship_type`, `orig_projected_additional_accrued_interest`, `payment_plan_start_date`
- String variables provided by borrowers. LendingClub categorized them into other variables: `desc`, `title`, and `emp_title`
- Loan settlement/charged off related variables -- wouldn't know them at the time of funding: `debt_settlement_flag`, `debt_settlement_flag_date`, `settlement_amount`, `settlement_date`, `settlement_percentage`, `settlement_status`, `settlement_term`, `recoveries`, `collection_recovery_fee`
- Loan payment related variables -- wouldn't know them at the time of funding: `funded_amnt`, `funded_amnt_inv`, `last_pymnt_amnt`, `out_prncp`, `out_prncp_inv`, `total_pymnt`, `total_pymnt_inv`, `total_rec_int`, `total_rec_late_fee`, `total_rec_prncp`, `last_pymnt_d`, `last_credit_pull_d`, `disbursement_method`
- `initial_list_status`: as the values "W' and "F" are randomly assigned
- Columns relating to co-borrowers. These columns only have non-null values from issue dates of 2017-03-01 and later (or 2015-10-01 and later for 'dti_joint', 'annual_inc_joint', and 'verification_status_joint'), and have around 97% missing values, so we remove these columns: `annual_inc_joint`, `dti_joint`, `revol_bal_joint`, `sec_app_chargeoff_within_12_mths`, `sec_app_collections_12_mths_ex_med`, `sec_app_earliest_cr_line`, `sec_app_inq_last_6mths`, `sec_app_mort_acc`, `sec_app_mths_since_last_major_derog`, `sec_app_num_rev_accts`, `sec_app_open_acc`, `sec_app_open_act_il`, `sec_app_revol_util`, `verification_status_joint`, `sec_app_fico_range_low`, `sec_app_fico_range_high`
- Other: `zip_code`

In [39]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier

In [40]:
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
sns.set(font_scale=1.5)
sns.set_style('ticks')

In [41]:
data = pd.read_csv('LC_15_18.csv').iloc[:, 1:]

## Data Cleaning

In [42]:
# check number of rows and cols in the dataset
data.shape

(927931, 73)

In [43]:
# change data type
data['issue_d'] = pd.to_datetime(data['issue_d'])
data['earliest_cr_line'] = pd.to_datetime(data['earliest_cr_line'])

In [44]:
# check missing values in the dataset
def missing_values_table(data):
    '''
    generate a table that contains attributes and its corresponing number of missing values 
    and percentage of missing value
    '''
    #Number of null values by column
    mv_df = pd.DataFrame(data.isna().sum(),columns=['Count_Missing'])
    
    #Portion of null values by column
    mv_df['Pctg_Missing'] = mv_df['Count_Missing']/len(data)

    #Sort by Missing_Count
    mv_df = mv_df.sort_values('Count_Missing',ascending=False)  
    
    return mv_df

mv_df = missing_values_table(data)

mv_df.head()

Count_Missing  Pctg_Missing
mths_since_last_record                 753505      0.812027
mths_since_recent_bc_dlq               696123      0.750188
mths_since_last_major_derog            659938      0.711193
mths_since_recent_revol_delinq         600515      0.647155
mths_since_last_delinq                 451425      0.486486

In [45]:
# Re-categorize emp_length into 4 categories
def process_emp_len(d):
    '''
    re_categorize the employment length in years to reduce the categories
    '''
    df = d.copy()
    # 0-1 years include null, < 1 year and 1 year
    df['emp_length'] = df['emp_length'].fillna('0-1 years')
    df['emp_length'] = np.where(df['emp_length'].isin([np.nan,None,'< 1 year','1 year']),'0-1 years',df['emp_length'])

    # 2-5 years include 2 years, 3 years, 4 years and 5 years
    df['emp_length'] = np.where(df['emp_length'].isin(['2 years','3 years','4 years', '5 years']),'2-5 years',df['emp_length'])

    # 6-9 years include 6 years, 7 years, 8 years and 9 years
    df['emp_length'] = np.where(df['emp_length'].isin(['6 years','7 years','8 years', '9 years']),'6-9 years',df['emp_length'])

    return df


# For missing values in columns that measure "months since": fill the nulls with maximum observed value +1
def process_month_since_cols(d):
    '''
    fill the null values with the maximum observed value + 1 in 'month since' related columns, 
    the borrowers who have never had delinquencies will have the largest value for number of months 
    since delinquency.
    '''
    df = d.copy()

    mon_since_cols = ['mo_sin_old_il_acct','mths_since_last_delinq','mths_since_last_major_derog',
                     'mths_since_last_record','mths_since_recent_bc_dlq','mths_since_recent_inq',
                     'mths_since_recent_revol_delinq','mo_sin_old_rev_tl_op','mths_since_recent_bc',
                     'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl']

    for i in mon_since_cols:
        df[i].fillna(df[i].max()+1, inplace=True)
    
    return df


# Change grade and sub_grade to numerical rankings (the lower the number, the less risky the loan)
def process_loan_grades(d):
    '''
    turn loan grade and sub_grade into ordinal mapping
    '''
    df = d.copy()

    # turn grade into ordinal mapping
    sorted_grades = sorted(d.grade.unique())
    grade_dict = dict(zip(sorted_grades, range(len(sorted_grades))))
    df['grade'] = df['grade'].map(grade_dict)

    # turn sub_grade into ordinal mapping
    sorted_subgrades = sorted(d.sub_grade.unique())
    subgrade_dict = dict(zip(sorted_subgrades, range(len(sorted_subgrades))))
    df['sub_grade'] = df['sub_grade'].map(subgrade_dict)

    return df


# Process the above transformation and a few more transformations
def process_loan_cols(d):
    df_processed = process_emp_len(d)
    df_processed = process_month_since_cols(df_processed)
    df_processed = process_loan_grades(df_processed)
    
    # add `credit_line_length` by compute the difference between the month the borrower's earliest reported credit line was opened
    # and the month which the loan was funded. Then discard the earliest_cr_line column.
    df_processed['credit_line_length'] = df_processed['issue_d'] - df_processed['earliest_cr_line']
    df_processed = df_processed.drop(['earliest_cr_line'], axis=1)

    # process `revol_util`, `int_rate`, `credit_line_length`
    df_processed['revol_util'] = df_processed['revol_util'].apply(lambda x: x/100)
    df_processed['int_rate'] = df_processed['int_rate'].apply(lambda x: x/100)
    df_processed['credit_line_length'] = df_processed['credit_line_length'].apply(lambda x: x.days)
    
    # generate `default` (binary variable, 0 means not defaulted, 1 means defaulted) and drop `loan_status`
    df_processed['default'] = df_processed['loan_status'].map({'Fully Paid': 0, 'Charged Off': 1, 'Default': 1,
                                                                  'In Grace Period': 1, 'Late (16-30 days)': 1, 
                                                                  'Late (31-120 days)': 1})
    df_processed = df_processed.drop(columns='loan_status')
    
    return df_processed


In [46]:
# create a new pre-process dataset
data_new = process_loan_cols(data)

In [47]:
# check missing values in new dataset
mv_df = missing_values_table(data_new) # the greatest missing percentage of a feature is around 0.01

# get the list of variables that have missing values
mv_list = mv_df[mv_df['Count_Missing'] > 0].index.tolist() # all the variables that have missing values are float

# impute these missing values with its median
for var in mv_list:
    data_new[var] = data_new[var].fillna((data_new[var].median()))

In [48]:
# check the data type for each attribute
types_data = pd.DataFrame(data_new.dtypes, columns=['Types'])

# check the categorical data
cat_var = types_data[types_data['Types'] == 'object'].index.tolist()
cat_var

['term',
 'emp_length',
 'home_ownership',
 'verification_status',
 'purpose',
 'addr_state',
 'application_type',
 'issue_month']

In [49]:
# create dummy variables for categorical variables
data_new = pd.get_dummies(data_new, columns=cat_var, drop_first=True) 

## Exploratory Data Analysis

In [50]:
data_new.groupby('default')['default'].count()/len(data_new)

default
0    0.756728
1    0.243272
Name: default, dtype: float64

75.67% of the loans are fully paid while 24.32% of the loans are defaulted.

### Feature Importance

Since there are more than 100 variables, we would like to first get a sense of which features are the most predictive by training a simple Random Forest Classifier and looking at the feature importance plot.

In [54]:
# When training the model, we need to first discard some columns:
# 'issue_d' has the date data type, which cannot be used in the rf model
# 'grade', 'sub_grade', and 'int_rate' are determined by loans' and borrowers' features and are directly associated with default risk
rf_data = data_new.drop(columns= ['issue_d', 'grade', 'sub_grade', 'int_rate'])
data_train, data_val = train_test_split(rf_data, test_size=.1, stratify=rf_data['default'], random_state=99)

X_train = data_train.drop(columns='default')
y_train = data_train['default']

X_val = data_val.drop(columns='default')
y_val = data_val['default']

In [55]:
rf_model = RandomForestClassifier(n_estimators=100, max_depth=None).fit(X_train, y_train)

importances = pd.DataFrame({'Columns':X_train.columns,'Feature_Importances':rf_model.feature_importances_})
importances = importances.sort_values(by='Feature_Importances',ascending=False)

In [57]:
importances.head(15)

Columns  Feature_Importances
16   last_fico_range_low             0.285693
15  last_fico_range_high             0.271585
60       term_ 60 months             0.015789
1            installment             0.014120
0              loan_amnt             0.013393
3                    dti             0.013067
58            issue_year             0.012789
25        bc_open_to_buy             0.010761
30  mo_sin_old_rev_tl_op             0.010492
24           avg_cur_bal             0.010405
59    credit_line_length             0.010386
54       tot_hi_cred_lim             0.010291
56        total_bc_limit             0.010199
2             annual_inc             0.010108
22      total_rev_hi_lim             0.010070

The 15 most important predictors are:
- `last_fico_range_low`: The lower boundary range the borrower’s last FICO pulled belongs to.
- `last_fico_range_high`: The upper boundary range the borrower’s last FICO pulled belongs to.
- `term`: The number of payments on the loan. Values are in months and can be either 36 or 60.
- `installment`: The monthly payment owed by the borrower if the loan originates.
- `loan_amnt`
- `dti`: A ratio calculated using the borrower’s total monthly debt payments on the total debt obligations, excluding mortgage and the requested LC loan, divided by the borrower’s self-reported monthly income.
- `issue_year`: 2015 to 2018.
- `bc_open_to_buy`: Total open to buy on revolving bankcards.
- `mo_sin_old_rev_tl_op`: Months since oldest revolving account opened.
- `avg_cur_bal`: Average current balance of all accounts.
- `credit_line_length`: The difference between the month the borrower's earliest reported credit line was opened and the month which the loan was funded.
- `tot_hi_cred_lim`: Total high credit/credit limit.
- `total_bc_limit`: Total bankcard high credit/credit limit.
- `annual_inc`: The self-reported annual income provided by the borrower during registration.
- `total_rev_hi_lim`: Total revolving high credit/credit limit.

Let's do some EDA on these 15 predictors.

In [72]:
# The two most important predictors
data_new.groupby('default').agg({
    'last_fico_range_low': 'median',
    'last_fico_range_high': 'median'
})

last_fico_range_low  last_fico_range_high
default                                           
0                      705.0                 709.0
1                      560.0                 564.0

Defaulted loan borrowers have both lower median lower boundary FICO and lower median higher boundary FICO than fully paid loan borrowers.

In [106]:
# change the 'term_ 60 months' variable to float to compute mean
data_new['term_ 60 months'] = data_new['term_ 60 months'].astype(np.float64)

# Predictors related to loans
data_new.groupby('default').agg({
    'term_ 60 months': 'mean',
    'installment': 'median',
    'loan_amnt': 'median'
})

term_ 60 months  installment  loan_amnt
default                                         
0               0.194299       364.70    12000.0
1               0.382887       408.22    14725.0

`term` is the number of payments on the loan. Values are in months and can be either 36 or 60. 38.29% of the defaulted loans are 60-months while only 19.43% of the fully paid loans are 60-months. 

Defaulted loans have a higher monthly payment owed by the borrower if the loan originates and a higher loan amount.

In [107]:
# change the 'term_ 60 months' variable back to uint8 data type
data_new['term_ 60 months'] = data_new['term_ 60 months'].astype(np.uint8)

In [108]:
# Predictors related to borrowers' financials
data_new.groupby('default').agg({
    'dti': 'median',
    'bc_open_to_buy': 'median',
    'mo_sin_old_rev_tl_op': 'median',
    'avg_cur_bal': 'median',
    'credit_line_length': 'median',
    'tot_hi_cred_lim': 'median',
    'total_bc_limit': 'median',
    'annual_inc': 'median',
    'total_rev_hi_lim': 'median'
})

dti  bc_open_to_buy  mo_sin_old_rev_tl_op  avg_cur_bal  \
default                                                             
0        17.50          5636.0                 167.0       8037.0   
1        20.07          4177.5                 153.0       5769.0   

         credit_line_length  tot_hi_cred_lim  total_bc_limit  annual_inc  \
default                                                                    
0                    5508.0         123653.0         16200.0     67000.0   
1                    5233.0          89510.0         13600.0     62000.0   

         total_rev_hi_lim  
default                    
0                 25500.0  
1                 22600.0

Borrowers of **defaulted** loans have:
- Higher debt-to-income ratio
- Lower total open to buy on revolving bankcards
- Less months since oldest revolving account opened
- Lower average current balance of all accounts
- Less difference between the month the borrower's earliest reported credit line was opened and the month which the loan was funded
- Lower total high credit/credit limit
- Lower total bankcard high credit/credit limit
- Lower median annual income
- Lower total revolving high credit/credit limit